core/js/src/object.ts

```diff
@@ -0,0 +1,181 @@
+  comment: {
+    text: string;
+  };
+  [AUDIT_SOURCE_FIELD]: Source;
+  [AUDIT_METADATA_FIELD]?: Record<string, unknown>;
+} & Omit<ParentExperimentIds | ParentProjectLogIds, "kind">;
+
+export type SanitizedExperimentLogPartialArgs = Partial<OtherExperimentLogFields> &
+  Partial<InputField>;
+
+export type ExperimentEvent = Partial<InputField> &
+  Partial<OtherExperimentLogFields> & {
+    id: string;
+    span_id?: string;
+    root_span_id?: string;
+    project_id: string;
+    experiment_id: string;
+    [IS_MERGE_FIELD]: boolean;
+  } & Partial<{
+    created: string;
+    span_parents: string[];
+    span_attributes: Record<string, unknown>;
+    context: Record<string, unknown>;
+    [PARENT_ID_FIELD]: string;
+    [AUDIT_SOURCE_FIELD]: Source;
+    [AUDIT_METADATA_FIELD]?: Record<string, unknown>;
+  }>;
+
+export type DatasetEvent = {
+  input?: unknown;
+  metadata?: unknown;
+  id: string;
+  project_id: string;
+  dataset_id: string;
+  created: string;
+} & ({ expected?: unknown } | { output?: unknown });
+
+export type LoggingEvent = Omit<ExperimentEvent, "experiment_id"> & {
+  org_id: string;
+  log_id: "g";
+};
+
+export type CommentEvent = IdField & {
+  created: string;
+  origin: {
+    id: string;
+  };
+  comment: {
+    text: string;
+  };
+  [AUDIT_SOURCE_FIELD]: Source;
+  [AUDIT_METADATA_FIELD]?: Record<string, unknown>;
+} & Omit<ParentExperimentIds | ParentProjectLogIds, "kind">;
+
+export type BackgroundLogEvent =
+  | ExperimentEvent
+  | DatasetEvent
+  | LoggingEvent
+  | CommentEvent;
+
+export const DefaultIsLegacyDataset = true;
+
+interface LegacyDatasetRecord {
+  id: string;
+  input: any;
+  output: any;
+  metadata: any;
+};
+
+interface NewDatasetRecord {
+  id: string;
+  input: any;
+  expected: any;
+  metadata: any;
+};
+
+export type DatasetRecord<IsLegacyDataset extends boolean = typeof DefaultIsLegacyDataset> =
+  IsLegacyDataset extends true ? LegacyDatasetRecord : NewDatasetRecord;
+
+export type AnyDatasetRecord = DatasetRecord<boolean>;
+
+export function ensureDatasetRecord<IsLegacyDataset extends boolean = typeof DefaultIsLegacyDataset>(r: AnyDatasetRecord, legacy: IsLegacyDataset): DatasetRecord<IsLegacyDataset> {
+  if (legacy) {
+    return ensureLegacyDatasetRecord(r) as DatasetRecord<IsLegacyDataset>;
+  } else {
+    return ensureNewDatasetRecord(r) as DatasetRecord<IsLegacyDataset>;
+  }
+}
+
+export function ensureLegacyDatasetRecord(r: AnyDatasetRecord): DatasetRecord<true> {
+  if ("output" in r) {
+    return r;
+  }
+  const row = {
+    ...r,
+    output: r.expected,
+  };
+  delete row.expected;
+  return row;
+}
+
+export function ensureNewDatasetRecord(r: AnyDatasetRecord): DatasetRecord<false> {
+  if ("expected" in r) {
+    return r;
+  }
+  const row = {
+    ...r,
+    expected: r.output,
+  };
+  delete row.output;
+  return row;
+}
+
+export function makeLegacyEvent(e: BackgroundLogEvent): BackgroundLogEvent {
+  if (!("dataset_id" in e) || !("expected" in e)) {
```

Comment:
I didn't do that because `ensureLegacyDatasetRecord` is typed to handle only dataset records, but these logged "events" also have `<object>_id`, `created`, and other stuff. I suppose there's some type foo I could do to share the functions, but not sure if worth doing

Reply:
ah gotcha. Yeah the way you have it is fine then.